samples/bluetooth/hci_rpmsg/src/main.c

```diff
@@ -0,0 +1,455 @@
+	static struct metal_io_region     *io;
+	static struct metal_device        *device;
+
+	/* Libmetal setup */
+	err = metal_init(&metal_params);
+	if (err) {
+		LOG_ERR("metal_init: failed - error code %d", err);
+		return err;
+	}
+
+	err = metal_register_generic_device(&shm_device);
+	if (err) {
+		LOG_ERR("Couldn't register shared memory device: %d", err);
+		return err;
+	}
+
+	err = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (err) {
+		LOG_ERR("metal_device_open failed: %d", err);
+		return err;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (!io) {
+		LOG_ERR("metal_device_io_region failed to get region");
+		return -ENODEV;
+	}
+
+	/* IPM setup */
+	ipm_tx_handle = device_get_binding("IPM_1");
+	if (!ipm_tx_handle) {
+		LOG_ERR("Could not get TX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_rx_handle = device_get_binding("IPM_0");
+	if (!ipm_rx_handle) {
+		LOG_ERR("Could not get RX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_register_callback(ipm_rx_handle, ipm_callback, NULL);
+
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[0]) {
+		LOG_ERR("virtqueue_allocate failed to alloc vq[0]");
+		return -ENOMEM;
+	}
+
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[1]) {
+		LOG_ERR("virtqueue_allocate failed to alloc vq[1]");
+		return -ENOMEM;
+	}
+
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.role = RPMSG_REMOTE;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	vdev.vrings_info = &rvrings[0];
+
+	/* setup rvdev */
+	err = rpmsg_init_vdev(&rvdev, &vdev, NULL, io, NULL);
+	if (err) {
+		LOG_ERR("rpmsg_init_vdev failed %d", err);
+		return err;
+	}
+
+	rdev = rpmsg_virtio_get_rpmsg_device(&rvdev);
+
+	err = rpmsg_create_ept(&ep, rdev, "k", RPMSG_ADDR_ANY,
+				  RPMSG_ADDR_ANY, endpoint_cb,
+				  rpmsg_service_unbind);
+	if (err) {
+		LOG_ERR("rpmsg_create_ept failed %d", err);
+		return err;
+	}
+
+	return err;
+}
+
+void main(void)
+{
+	int err;
+
+	/* incoming events and data from the controller */
+	static K_FIFO_DEFINE(rx_queue);
+
+	/* initialize RPMSG */
+	err = hci_rpmsg_init();
+	if (err != 0) {
+		return;
+	}
+
+	LOG_DBG("Start");
+
+	/* Enable the raw interface, this will in turn open the HCI driver */
+	bt_enable_raw(&rx_queue);
+
+	/* Spawn the TX thread and start feeding commands and data to the
+	 * controller
+	 */
+	k_thread_create(&tx_thread_data, tx_thread_stack,
+			K_THREAD_STACK_SIZEOF(tx_thread_stack), tx_thread,
+			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+	while (1) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		err = hci_rpmsg_send(buf);
+		if (err) {
+			LOG_ERR("Failed to send");
```

Comment:
To help debugging, perhaps make this `"Failed to send (err %d)", err`

Reply:
done